### PR TITLE
Refs [TASK][26276] Remote Data Source

### PR DESCRIPTION
**Pull request by toandv98**
## Related Tickets

- [#TASK 26276](https://edu-redmine.sun-asterisk.vn/issues/26276)

## WHAT

- Create Http request API, create Remote Data Source.

## HOW

- Create Http request API using HttpUrlConnection, create Remote Data Source.

## Review Checklist

| Category      | View Point                                                   | Description                                                  | Expected Reviewer Answer | Reviewer1 (ToanDV) | Reviewer2 (HuyTQ)   |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------ | ------------------ | ------------------ |
| General       | Code should be self-explanatory.                             | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Is this function or class too big? If yes, is the function or class having too many responsibilities? |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| General       | Does the code follow DRY?: Do not Repeat Yourself            | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Exception handling and cleanup (dispose) resources           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| General       | Does the code build and execute correctly?                   |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the source code have end-line errors?                   |                                                              | NO                       | <li>- [x] no</li>  | <li>- [x] no</li>  |
| Conventions   | No hard coding, use constants/configuration values           |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Conventions   | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Redmine       | Does the ticket follow Sun* Redmine working process?         | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |
| Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |                                                              | YES                      | <li>- [x] yes</li> | <li>- [x] yes</li> |

## Notes (Optional)

*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        return diskIO
+    }
+
+    fun networkIO(): Executor {
+        return networkIO
+    }
+
+    fun mainThread(): Executor {
+        return mainThread
+    }
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt]:
```suggestion
    val diskIO: Executor,
    val networkIO: Executor,
    val mainThread: Executor
) {

```

File: app/src/main/java/com/edu/sun/oereminder/data/source/remote/model/Member.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.edu.sun.oereminder.data.source.remote.model
+
+data class Member(
+    val accountId: Int,
+    val avatarImageUrl: String,
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/data/source/remote/model/Member.kt]:
```suggestion
    val avatarUrl: String,
```
avatar nó đã là ảnh r nên k cần Image đâu ô =))

File: app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        val INSTANCE = AppExecutors(
+            Executors.newSingleThreadExecutor(),
+            Executors.newFixedThreadPool(NUM_FIXED_THREAD),
+            MainThreadExecutor()
+        )
+    }
+
+    companion object {
+        @JvmStatic
+        fun getInstance(): AppExecutors {
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt]:
```suggestion
        fun getInstance() = Holder.INSTANCE
```

File: app/src/main/java/com/edu/sun/oereminder/data/repository/MessageRepository.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import com.edu.sun.oereminder.data.SourceCallBack
+import com.edu.sun.oereminder.data.source.remote.model.Message
+
+interface MessageRepository {
+
+    fun getMessages(callBack: SourceCallBack<List<Message>>)
+
+    fun sendMessage(message: String)
+
+    fun updateMessage(messageId: Int, message: String)
```
Comment (thonc-2135) [inline app/src/main/java/com/edu/sun/oereminder/data/repository/MessageRepository.kt]: góp ý chút về cách đặt tên , hàm ô đặt là updateMeassage nhưng biến truyền vào lại có message, như v là update Message hay là update message, t nghĩ nên đổi lại thành content hay gì đó rõ hơn để có thể clean hơn

File: app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+
+    private object Holder {
+        val INSTANCE = AppExecutors(
+            Executors.newSingleThreadExecutor(),
+            Executors.newFixedThreadPool(NUM_FIXED_THREAD),
+            MainThreadExecutor()
+        )
+    }
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/AppExecutors.kt]: cho luôn vào companion object đi em

File: app/src/main/java/com/edu/sun/oereminder/data/SourceCallBack.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.edu.sun.oereminder.data
+
+interface SourceCallBack<T> {
+    fun onFinish(data: T)
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/data/SourceCallBack.kt]:
```suggestion
    fun onSuccess(data: T)
```

File: app/src/main/java/com/edu/sun/oereminder/data/source/prefs/PrefsHelperImpl.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        encryptedPrefs = context.encryptedPrefs(ENCRYPTED_PREFS_NAME)
+        defaultPrefs = context.defaultPrefs()
+    }
+
+    override fun getApiToken() = encryptedPrefs?.get(KEY_API_TOKEN, "") ?: ""
+
+    override fun getRoomId() = defaultPrefs?.get(KEY_ROOM_ID, "") ?: ""
+
+    companion object {
+        @Volatile
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/data/source/prefs/PrefsHelperImpl.kt]: Question: Annotation này có tác dụng gì thế em?

File: app/src/main/java/com/edu/sun/oereminder/data/source/remote/model/Plan.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.edu.sun.oereminder.data.source.remote.model
+
+data class Plan(var messId: Int, var date: String, var plan: String) {
```
Comment (datdt-1935) [inline app/src/main/java/com/edu/sun/oereminder/data/source/remote/model/Plan.kt]: break line

**Review by datdt-1935 — APPROVED** [review]

File: app/src/main/java/com/edu/sun/oereminder/data/source/prefs/PrefsHelperImpl.kt (hunk 7 above)
Comment (toandv98) [reply app/src/main/java/com/edu/sun/oereminder/data/source/prefs/PrefsHelperImpl.kt]: @Volatitle nó giống từ khóa volatile trong java, khi hai thread khác nhau cùng truy cập thì dữ liệu sẽ được update ngay cho thread còn lại ạ.

**Review by thonc-2135 — APPROVED** [review]

**Review by huytq-2422 — APPROVED** [review]

**Pull request merged**